bug fix/1747. enable error logging for invalid flags

ethanwater opened the pull request:
## Related Issue
  - fixes bug https://github.com/keploy/keploy/issues/1747

Closes: #1747 

#### Describe the changes you've made

- Remove the silencing of Record.
- Implement proper handling of ValidateFlags() calls.

## Type of change

- [x] My code follows the code style of this project.
- [x ] Bug fix (non-breaking change which fixes an issue)
- [x ] New feature (non-breaking change which adds functionality)
- [x ] Code style update (formatting, local variables)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

ethanwater commented:
I have read the CLA Document and I hereby sign the CLA

ethanwater commented:
recheck

shivamsouravjha commented:
do these fixes work, I tried the same them locally didn't work for me 
<img width="779" alt="image" src="https://github.com/keploy/keploy/assets/60891544/42cc1635-c896-46e8-81cd-24aeba668ea1">


if possible attach the screenshots

ethanwater commented:
@shivamsouravjha , using your example it works.

<img width="1272" alt="Screenshot 2024-03-29 at 11 08 44" src="https://github.com/keploy/keploy/assets/124078937/7b9edcc8-421f-48d9-823c-7a4d6498e1c4">

shivamsouravjha commented on cli/config.go:
```diff
@@ -24,10 +24,13 @@ func Config(ctx context.Context, logger *zap.Logger, cfg *config.Config, service
 		Example: "keploy config --generate --path /path/to/localdir",
 		PreRunE: func(cmd *cobra.Command, _ []string) error {
-			return cmdConfigurator.ValidateFlags(ctx, cmd)
+			if err := cmdConfigurator.ValidateFlags(ctx, cmd); err != nil {
+				utils.LogError(logger, err, "failed to validate flags")
+				return err
+			}
+			return nil
```
what's the use of logerror?

ethanwater replied:
This is proper use of handling errors according to Google's Golang documentation. The purpose of it is to simply log errors following production standard.

ethanwater replied:
https://go.dev/blog/error-handling-and-go

shivamsouravjha replied:
I appreciate the enthusiasm but the said logs aren't used. I put a print function  in utils.LogError and it doesn't work.

I understand its a official method of doings things but it can be cross checked if its the code snippet is working or not.



EXAMPLE: 
<img width="914" alt="image" src="https://github.com/keploy/keploy/assets/60891544/2a6c1dbf-dc06-48b8-879b-ad833e7bccf1">
****

ethanwater replied:
The LogError is a function that is used throughout the entirety of the codebase and was designed by the developers.

**What you're looking at is outside the scope of the issue you posted. The but has been fixed and is ready to be merged. **

Also to trigger that, you would need to make ValidateFlags fail.

If you have further questions about LogError() Ask the creators of Keploy themselves.

ethanwater replied:
Perhaps @slayerjain can provide some insight on the LogError function. But yes, the function the original developers implemented does work in the proper circumstances of error handling and detection.

slayerjain replied:
@ethanwater @shivamsouravjha its totally cool to use the `LogError` wrapper. It's something we recently added to avoid logging unnecessary errors while ctx is cancelled. We might add more validations in the future.

ethanwater replied:
Thank you @slayerjain for the review. I can implement these features. Should I implement these in the current PR or another one after merge?

slayerjain replied:
I would really appreciate implementing in the current PR. Happy to actively share my thoughts if you need :)

ethanwater replied:
@slayerjain Awesome, looking at it- I agree with these features. What do you think of this:

- ❌ and red text used for fatal errors/ return err
- 📌 or ⚠️ and yellow text used for info logs/warnings

ethanwater commented:
@slayerjain I implemented the features you requested! I also removed the redundant logging as the previous logging is already substantial enough. 

Note: I set the revamped error output in the AddFlags function due to the requirement of the command.SetFlagErrorFunc() being called post initialization. I have tried to search for any work arounds but they were all very obscure and overly complex. I prefer clean and fast code over dirty and complex code!

slayerjain commented on cli/provider/cmd.go:
```diff
@@ -153,6 +154,15 @@ func NewCmdConfigurator(logger *zap.Logger, config *config.Config) *CmdConfigura
 func (c *CmdConfigurator) AddFlags(cmd *cobra.Command) error {
+	//sets the displayment of flag-related errors
+	cmd.SilenceErrors = true
+	cmd.SetFlagErrorFunc(func(cmd *cobra.Command, err error) error {
```
@ethanwater This is very nice!

slayerjain commented on cli/config.go:
```diff
@@ -24,10 +24,12 @@ func Config(ctx context.Context, logger *zap.Logger, cfg *config.Config, service
 		Example: "keploy config --generate --path /path/to/localdir",
 		PreRunE: func(cmd *cobra.Command, _ []string) error {
-			return cmdConfigurator.ValidateFlags(ctx, cmd)
+			if err := cmdConfigurator.ValidateFlags(ctx, cmd); err != nil {
```
since we are not doing anything here, it can be left unchanged to `return cmdConfigurator.ValidateFlags(ctx, cmd)` as in the previous commit.

slayerjain reviewed (changes requested):
This looks awesome Please just remove the redundant error check in the 3 commands as pointed by the linter too and the redundant `cmd` variable. We should be good to roll after that :)

slayerjain commented on cli/provider/cmd.go:
```diff
@@ -153,6 +154,15 @@ func NewCmdConfigurator(logger *zap.Logger, config *config.Config) *CmdConfigura
+	//sets the displayment of flag-related errors
+	cmd.SilenceErrors = true
+	cmd.SetFlagErrorFunc(func(cmd *cobra.Command, err error) error {
+		color.Red(fmt.Sprintf("❌ error: %v", err))
```
@ethanwater This is very nice! I really like the output now.

ethanwater commented:
@slayerjain Thanks again for the review! And I'm glad you like the new output! I have applied the changes you recommended.

slayerjain commented:
@ethanwater Thanks for the contribution and patience. Looking forward to your next contribution :)
